src/server/connection.py

```diff
@@ -120,31 +126,53 @@ def __str__(self) -> str:
 
 
 class PeerConnections:
-    def __init__(self, all_connections: List[ChiaConnection] = []):
+    def __init__(
+        self, local_type: NodeType, all_connections: List[ChiaConnection] = []
+    ):
         self._all_connections = all_connections
-        # Only full node peers are added to `peers`
-        self.peers = Peers()
-        for c in all_connections:
-            if c.connection_type == NodeType.FULL_NODE:
-                self.peers.add(c.get_peer_info())
+        self.local_type = local_type
+        self.introducer_peers = None
+        self.connection = None
+        if local_type == NodeType.INTRODUCER:
+            self.introducer_peers = IntroducerPeers()
+            for c in all_connections:
+                if c.connection_type == NodeType.FULL_NODE:
+                    self.introducer_peers.add(c.get_peer_info())
         self.state_changed_callback: Optional[Callable] = None
+        self.full_node_peers_callback: Optional[Callable] = None
+        self.wallet_callback: Optional[Callable] = None
+        self.max_inbound_count = 0
 
     def set_state_changed_callback(self, callback: Callable):
         self.state_changed_callback = callback
 
+    def set_full_node_peers_callback(self, callback: Callable):
+        self.full_node_peers_callback = callback
+
+    def set_wallet_callback(self, callback: Callable):
+        self.wallet_callback = callback
+
     def _state_changed(self, state: str):
         if self.state_changed_callback is not None:
             self.state_changed_callback(state)
 
     def add(self, connection: ChiaConnection) -> bool:
+        if not connection.is_outbound:
+            if (
+                connection.connection_type is not None
+                and not self.accept_inbound_connections(connection.connection_type)
+            ):
+                raise ProtocolError(Err.MAX_INBOUND_CONNECTIONS_REACHED)
+
         for c in self._all_connections:
             if c.node_id == connection.node_id:
-                return False
+                raise ProtocolError(Err.DUPLICATE_CONNECTION, [False])
```

Comment:
What's the point of this?

Reply:
same as before, not to add duplicate connections and fail on handshake. this just got moved to .add to free up handshake code.